File: 树/树的层次遍历-40/103.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root: return []
        dq, res, flag = deque(), [], 1
        dq.append(root)
        while len(dq) != 0:
            temp= []
            for i in range(len(dq)):
                node = dq.popleft()
                temp.append(node.val)
                if node.left: dq.append(node.left)
                if node.right: dq.append(node.right)
            if flag == 1:
                res.append(temp[:])
                flag = 0
            else:
                res.append(temp[::-1])
                flag = 1
        return res

File: 树/树的层次遍历-40/103.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> ans = new LinkedList<List<Integer>>();
        if (root == null) {
            return ans;
        }
        boolean isLeft = true;
        Queue<TreeNode> que = new LinkedList<>();
        que.add(root);
        while (!que.isEmpty()) {
            Deque<Integer> levelList = new LinkedList<Integer>();
            int n = que.size();
            for (int i = 0; i < n; i++) {
                TreeNode cur = que.poll();
                if (isLeft) {
                    levelList.addLast(cur.val);                    
                }
                else {
                    levelList.addFirst(cur.val);
                }
                if (cur.left != null) {
                    que.add(cur.left);
                }
                if (cur.right != null) {
                    que.add(cur.right);
                }
            }
            isLeft = !isLeft;
            ans.add(new LinkedList<Integer>(levelList));
        }
        return ans;
    }
}